### ISSUE TITLE
Handle type conversions in SQL generation

### ISSUE THREAD

**Issue by ajcvickers** [aspnet/EntityFrameworkCore#10434]
Consider cases like #10427 which involves SQL translations for DateTime calculations. If the property has been mapped to some non-datetime type, then this kind of translation will not work anymore. Instead, query should either:
* Do a different translation
* Revert to client evaluation

Probably client eval is what we can realistically do in the short term, but in the future it would be good to have the type mapping/value conversion participate in generating the translations.

**Comment by ajcvickers** [comment]
Triage: client eval for now.

**Comment by ajcvickers** [comment]
Triage: For 2.1, we will:
* Detect places where query is going to use a converted type in a place "that matters"--for example, Where clause--and output a warning.
* The warning will be for every operation other than equality/inequality
* The warning will direct people to consider client eval

For 2.1, we will not:
* Force client eval in these cases--for example, forcing client eval on all ordering would be bad since often only a stable ordering is needed and forcing client eval would kill perf
* Try to add API around this area since there are many considerations and there is not enough time to bake it for 2.1. See #10265

**Comment by ajcvickers** [comment]
See also #11156

**Comment by anpete** [comment]
Warning is done. Clearing milestone to track further improvements.

**Comment by michaldudak** [comment]
What about equality/inequality in Where clause?
It seems that when there is a converted type in a Where clause (something like `.Where(b => b.BlogId == new BlogId("8807e48e-1236-4ef0-a8c9-3de730e52e1b")` in my case, where `BlogId` is a struct with a type conversion to/from a Guid) the client-side evaluation kicks in. So in this case, all Blogs would get fetched from the database, which is far from acceptable.

**Comment by bricelam** [comment]
I actually don't think this is necessary, but we'll pull in parts of it if needed

**Comment by divega** [comment]
Thanks @bricelam. I think I was confusing this issue with the general ability to push a type conversion down to SQL, e.g. in the case of spatial for SQL Server, to get WKT or WKB, so that type like sqlgeography and sqlgeometry that cannot be represented in Core CLR are not present in the results. Do you know if we are tracking this elsewhere?

**Comment by ajcvickers** [comment]
@divega I think #10861 is the issue you were looking for. It's currently in backlog.

**Comment by MovGP0** [comment]
I found that in my codebases there are only a few use cases that EF Core can't handle. Most importantly are type aliases and time conversions (NodaTime). 

### Type Alias

A type alias (there might be a better name for it) is something that wraps a scalar primitive type like a `string`, `int`, `guid`, or something similar in a `struct`: This is mosly done for type savety when writing a LINQ query, which can drastically reduce bugs in an large codebase. 

You can do something like that easlily in languages like F#, where you cas use union types to do the job: 

```fsharp
type OrderId = OrderId of System.Guid;
type CustomerId = CustomerId of System.Guid;
```

This prevents you from using an `CustomerId` where one would need an `OrderId` and vice-versa. Unfortunately, there is no such thing in C#, and I am unsure how well the support by EF is for F#, so one must rely on custom structs. 

#### Example: Custom Struct

Instead of using a plain `Guid` type for an order's id, you could write something like: 

```csharp
    public struct OrderId : IEquatable<OrderId>, IComparable<OrderId>
    {
        private Guid Id { get; }

        public OrderId(Guid id)
        {
            Id = id;
        }

		public OrderId(string id)
        {
            Id = new Guid(id);
        }

        public static explicit operator Guid(OrderId id)
        {
            return id.Id;
        }

        public static explicit operator OrderId(Guid id)
        {
            return new OrderId(id);
        }

	public bool Equals(OrderId other)
        {
            return Id.Equals(other.Id);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            return obj is OrderId id && Equals(id);
        }

        public override int GetHashCode()
        {
            return Id.GetHashCode();
        }

	public static bool operator ==(OrderId x, OrderId y)
        {
            return x.Id == y.Id;
        }

        public static bool operator !=(OrderId x, OrderId y)
        {
            return !(x == y);
        }

	public int CompareTo(OrderId other)
	{
		return Id.CompareTo(other.Id);
	}

	public override string ToString()
	{
		return Id.ToString();
	}

	public static OrderId Empty => new OrderId(Guid.Empty);
    }
```

### Example Use-Case 

Now lets have a class like the following: 

```csharp
public sealed class Order
{
    public OrderId Id { get; set; }

    // Navigational property for n:m relations
    public CustomerId CustomerId { get; set; }
    public Customer Customer { get; set; }

    // Navigational property gor 1:n relations 
    public ICollection<Item> Items { get; set; } = new HashSet<Item>();

    public NodaTime.Instant OrderDate { get; set; }
    public NodaTime.Instant? DeliveryDate { get; set; }

    // additional properties... 
}
```

Also assume that I did implement ValueConverters and register those with the ModelBulder: 

```csharp
var entity = modelBuilder.Entity<Order>();

entity.ToTable("Order");
entity.HasKey(e => e.Id);

entity.Property(e => e.Id)
    .HasColumnName("id")
    .HasColumnType("uniqueidentifier")
    .IsRequired()
    .HasConversion<OrderId>(id => (OrderId)id, oid => (Guid)oid);

/* other properties */
```

So now I want to do something like getting the order details of a set of selected orders within a time range: 

```csharp
// using an injected NodaTime IClock to get the current time 
var now = Clock.GetCurrentInstant();
var threeDaysAgo = now.Minus(Duration.FromDays(3));

var result = await Context.Orders
    .Where(order => order.OrderDate >= threeDaysAgo) // doesn't work
    .Where(order => order.DeliveryDate == null || order.DeliveryDate > threeDaysAgo) // doesn't work
    .Where(order => orderIds.Contains(order.Id)) // doesn't work
    .Where(order => nullableCustomerId == null || order.CustomerId == nullableCustomerId) // doesn't work
    .Include(order => order.Items)
    .Select(order => new OrderDetail { OrderId = order.Id, /* ... */ })
    .AsNoTracking()
    .ToListAsync(cancellationToken)
    .ConfigureAwait(false);
```

### Remarks

I guess it would be possible to write custom extension methods for `IQueryable<T>` which build the proper expression tree. I am just not sure how that would work. So any hint would be very welcome. 

### References

* [Designing with types: Single case union types](https://fsharpforfunandprofit.com/posts/designing-with-types-single-case-dus/)
* [How to: Use Expression Trees to Build Dynamic Queries (C#)](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/expression-trees/how-to-use-expression-trees-to-build-dynamic-queries)

**Comment by ajcvickers** [comment]
@MovGP0 For the cases where you comment "doesn't work" do you get an exception, bad SQL, or does it fail in some other way? Can you please post details of the stracktraces, bad SQL, or other failures?

**Comment by MovGP0** [comment]
@ajcvickers those cases work in memory and not in the SQL Server. 

If I get an exception or not depends in the configuration. By default it's just a warning and very slow.

**Comment by MovGP0** [comment]
I guess that there might be the need to register type conversions globally in the `DbContext`, such that the expression tree builder in the LINQ clauses can access it  (ie. by trying to cast the  `IQueryable` object to some new  `IConversionProvider` base type) and generate the proper expression tree.

**Comment by NickCraver** [comment]
Not sure if related, but we're hitting similar, a simple example of:

```c#
var u = DB.Users.Single(u => u.GUID == new Guid(guidString));
```
...won't translate, where as:
```c#
var guid = new Guid(guidString);
var u = DB.Users.Single(u => u.GUID == guid);
```
...works fine. This was a regression from Linq2SQL where it was a non-issue and translated to a server-side SQL query.

**Comment by smitpatel** [comment]
@NickCraver - It is slightly unrelated to this issue but on my radar. I filed #13549 for tracking & visibility.

**Comment by divega** [comment]
@MovGP0 FYI, the limitation you describe is different from the one covered by this issue. We are tracking it in https://github.com/aspnet/EntityFrameworkCore/issues/12045.

**Comment by smitpatel** [comment]
The SQL translator would have access to type mapping of each translated fragment, allowing it to inspect the type mapping and call appropriate function on server side.

**Comment by apavelm** [comment]
@ajcvickers
Is any progress on this?

I'm using NodaTime converters, but still unable to execute filtering/sorting on the server side.

The latest version of EF Core.

I see a warning, but cannot filter resultset by NodaTime's LocalDate/LocalDateTime/Instant 

`The LINQ expression 'where ([s].StartTime >= __dateStart_1)' could not be translated and will be evaluated locally.`

Are any plans to implement this in further releases, or at least ability how to help SqlTranslator to do that?

I tried to add SqlColumnType for an EntityTypeBuilder
`config.Property(t => t.StartTime).HasColumnType("datetime").HasConversion(ConverterHelpers.LocalDateTimeConverter);`

`
public static ValueConverter LocalDateTimeConverter =>
            new ValueConverter<LocalDateTime, DateTime>(v => v.ToDateTimeUnspecified(), v => LocalDateTime.FromDateTime(v));
`
But it doesn't work. Any ETA on this?

Thank you.

**Comment by ajcvickers** [comment]
@apavelm This is in the 3.0 milestone, which means we still plan to look at it for EF Core 3.0. I don't expect every case of this to be fixed for that release, but I do expect we will make progress.

**Comment by apavelm** [comment]
@ajcvickers Thank you for info!

**Comment by ajcvickers** [comment]
Notes from triage:
* Cases where the representation in the database has not changed (e.g. some type that is not `DateTime`, but still translated to `datetime` in the database) should work for 3.0 with the query pipeline changes
* Cases where the representation is different (e.g. `DateTime` to strings in the database) is dependent on #15979 and may be punted. depending on what we see from feedback and testing.

**Comment by ajcvickers** [comment]
See also scenario in #16639

**Comment by ajcvickers** [comment]
See also the case here: #17879

**Comment by OskarKlintrot** [comment]
Will this make it to 5.0? Would love to be able to combine NodaTime with EF Core + SQL Server!

**Comment by ajcvickers** [comment]
@OskarKlintrot We're not currently scheduling this for 5.0. However, I will discuss with team. I don't expect we can fit it in, but we may make it a stretch goal.

**Comment by ajcvickers** [comment]
Update from triage: we will look at this as a stretch goal for 5.0. Specifically, we may be able to identify some relatively cheap things we can do to enable common scenarios without the API becoming blocking for later work on more advanced scenarios.

**Comment by ajcvickers** [comment]
See also case here: #20094

**Comment by Lobosque** [comment]
Another use case that might be worth sharing: https://github.com/Lobosque/ef-value-object-poc

**Comment by ajcvickers** [comment]
See example in #21956

**Comment by ajcvickers** [comment]
See https://github.com/dotnet/efcore/issues/10434

**Comment by AndriySvyryd** [comment]
> See #10434

This **is** #10434

**Comment by OpenSpacesAndPlaces** [comment]
Ran into this running contains on a Int32[] that is value converting from a string field.
`Translation of method 'System.Linq.Enumerable.Contains' failed.`

**Comment by AndriySvyryd** [comment]
A design proposal is to have two store-side value converters on the type mapping that wouldn't impact the store type used for migrations:

- One for serialization/deserialization, it can affect the provider type
- One for query when using type inference. Composition can be used to construct the appropriate converter chain depending on the context (e.g. when used as an argument of a function that expects a different store type)

**Comment by voroninp** [comment]
Will it be planned for EF 9?

**Comment by ajcvickers** [comment]
@voroninp Too early to say yet, but I think it's probably quite unlikely that we'll fully implement this in EF Core 9.

**Comment by voroninp** [comment]
@ajcvickers , if I wanted to extend that part, where should I look at? 

In most cases I need translation of comparison operators (`>`, `>=`, `<`, `<=`) and in most cases it's just the same operation applied to the converted value.

**Comment by roji** [comment]
@voroninp you'd have to provide more context on exactly which (converted) types you want to add comparison operators to.

**Comment by voroninp** [comment]
For example we have ints which are time HHmmss, we convert them to our own type `RadioTime` (don't ask why not `TimeOnly`) which is effectively just a wrapper. And we need queries where we compare times.

**Comment by roji** [comment]
You'll have to define static functions representing operations over RadioTime (e.g. GreaterThan), and then use [user-defined function mapping](https://learn.microsoft.com/en-us/ef/core/querying/user-defined-function-mapping) to map them to the SQL operators.

**Comment by voroninp** [comment]
@roji

> System.InvalidOperationException Message=The parameter 'left' for the DbFunction DateIndex.op_GreaterThan(DateIndex, DateIndex)' has an invalid type 'DateIndex'. Ensure the parameter type can be mapped by the current provider.

**Comment by voroninp** [comment]
@roji, I need some help:

```CSharp
modelBuilder.HasDbFunction(
    typeof(SqlTranslation).GetMethod(
        nameof(SqlTranslation.LessThan), 
        BindingFlags.Public | BindingFlags.Static, 
        new[] { typeof(DateTimeOffset), typeof(DateTimeOffset) })!)
    .HasTranslation(
        args => new SqlBinaryExpression(
            ExpressionType.LessThan,
            args[0],
            args[1],
            typeof(bool),
            new BoolTypeMapping("boolean", DbType.Boolean)));
```

This works, but if I replace with `typeof(DateTimeOffset).GetMethod("op_LessThan", BindingFlags.Public | BindingFlags.Static)!` it fails. Are special methods not checked for custom mapping?

**Comment by roji** [comment]
@voroninp can you please open a separate issue with the precise repro code sample that shows what additional check you're proposing?
